### PR TITLE
Give stampinf correct DriverVer information

### PR DESCRIPTION
**Pull request by jkunkee**
Give stampinf correct DriverVer information

VS2017 has some nice INF-processing infrastructure, but it uses stampinf
which insists on adding the DriverVer directive to the INF file itself.
This overwrites whatever the INF template produces.

This commit plumbs the INF template values in through stampinf.

Note that the VS2017 GUI field "Driver Version Number" is stored in the
project file field `TimeStamp`.

### CODE REVIEW

File: src/tap-windows6.vcxproj.in [hunk 1]
```diff
@@ -121,6 +121,10 @@
     <Link>
       <AdditionalDependencies>%(AdditionalDependencies);$(KernelBufferOverflowLib);$(DDK_LIB_PATH)ntoskrnl.lib;$(DDK_LIB_PATH)hal.lib;$(DDK_LIB_PATH)wmilib.lib;ndis.lib;ntstrsafe.lib;wdmsec.lib</AdditionalDependencies>
     </Link>
+    <Inf>
+      <TimeStamp>@PRODUCT_TAP_WIN_MAJOR@.00.00.@PRODUCT_TAP_WIN_MINOR@</TimeStamp>
+      <DateStamp>@PRODUCT_TAP_WIN_RELDATE@</DateStamp>
```
Comment (selvanair) [inline src/tap-windows6.vcxproj.in]: That format was before my version number patch (PR #48) -- this is internally merged in @mattock's test executable. So could you please change the TimeStamp definition to 
@PRODUCT_TAP_WIN_MAJOR@.@PRODUCT_TAP_WIN_MINOR@.@PRODUCT_TAP_WIN_REVISION@.@PRODUCT_TAP_WIN_BUILD@

**Comment by jkunkee** [comment]
Yes; working on it now.

File: src/tap-windows6.vcxproj.in (hunk 1 above)
Comment (jkunkee) [reply src/tap-windows6.vcxproj.in]: Done.

**Comment by selvanair** [comment]
@mattock  Can we retest with this patch included? This will give a minor merge conflict in OemVista.inf.in. Easily resolved by just accepting "DriverVer=filled in by stampinf" to replace the line coming from PR 48.

**Comment by jkunkee** [comment]
Note that it doesn't matter what's on that line, so keeping it either way is technically valid (though one is more explanatory).

**Comment by selvanair** [comment]
Yes, what is added here should be preferred as it documents that date and version will get stamped by stampinf.

**Comment by cron2** [comment]
Looks promising...!  Thanks, Jon.

**Comment by selvanair** [comment]
Forgot to mention that I did a rebuild including this patch and checked that the inf processed by stampinf  now gets the right date and version as

DriverVer = 07/26/2018,9.22.2.601

By the way, the new tapinstall.exe has acquired a dependency on VCRUNTIME140.dll which  cannot be expected to be present.

**Comment by jkunkee** [comment]
Irritating. The CRT can be statically linked, but it may require changing the project. I’ll take a look.

Sent from my Windows 10 device
[15254 - 16.0.10325.20094]

**Comment by jkunkee** [comment]
There are a few options:

1. We can fork devcon.exe and then it's easy.
2. We can suggest the change to static linking upstream. I don't know how that might go.
3. We can attempt command line hackery to force static linking.

I'm working on 3 at the moment.

**Comment by cron2** [comment]
Hi,

On Tue, Jul 31, 2018 at 08:50:30PM +0000, Jon Kunkee wrote:
> There are a few options:
> 
> 1. We can fork devcon.exe and then it's easy.
> 2. We can suggest the change to static linking upstream. I don't know how that might go.
> 3. We can attempt command line hackery to force static linking.
> 
> I'm working on 3 at the moment.

Can we bundle this DLL?

gert
-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             gert@greenie.muc.de

**Comment by jkunkee** [comment]
The SDK comes with the VC 2017 Redist intended for exactly that purpose, yes. I am fuzzy on the fine print here, but:

* an installer for it is under `%root_of_EWDK%\Program Files\Microsoft Visual Studio\2017\BuildTools\VC\Redist\MSVC` as vcredist_x86.exe
* The DLL itself is at `%root_of_EWDK%\Program Files\Microsoft Visual Studio\2017\BuildTools\VC\Redist\MSVC\14.13.26020\x86\Microsoft.VC141.CRT` and is likely to be standalone and downlevel-compatible to Windows 7.

**Comment by jkunkee** [comment]
#3 does not look practical.

**Comment by selvanair** [comment]
Are we allowed to redistribute this runtime or does it have to be downloaded on the target host at install time? If we can't bundle it, we may have to revert back to the old prebuilt binaries until a good solution can be found.

**Comment by jkunkee** [comment]
Yes: https://docs.microsoft.com/en-us/cpp/ide/redistributing-visual-cpp-files

Sorry I’m not able to do more explication at the present moment.

Sent from my Windows 10 device
[15254 - 16.0.10325.20094]

**Comment by rozmansi** [comment]
AFAIK, you are not allowed to distribute Microsoft's run-time DLLs along your binaries. I remember reading the EULA a few years back, explaining you MUST use either vcredist_x86.exe or MSM merge modules to install the VC run-time libraries.

I suggest you stick to the old-reliable version of devcon.exe/tapinstall.exe for the time being. I have MSI packaging for TAP driver finished. MSI packaging for OpenVPN itself should follow soon. None will need devcon.exe.

**Comment by selvanair** [comment]
> AFAIK, you are not allowed to distribute Microsoft's run-time DLLs along your binaries. I remember reading the EULA a few years back, explaining you MUST use either vcredist_x86.exe or MSM merge modules to deploy the VC run-time libraries.

This used to be my understanding too. But as Jon pointed out and based on deployment methods listed [here](https://docs.microsoft.com/en-us/cpp/ide/choosing-a-deployment-method) copying the dlls to the applications's directory on target computer is a possible option though not recommended.

<quote>

_Local deployment, in which you copy particular Visual C++ DLLs from your Visual Studio installation—typically in \Program Files (x86)\Microsoft Visual Studio version\VC\Redist\platform\library\—and install them on target computers in the same folder as the application executable._ 

</quote>

Note that files in [VSFOLDER]\VC\Redist are expressly listed in as redistributable (unmodified) except for 3 exceptions (named with a prefix nonredist)

From an end user's perspective using vcredist_x86.exe or merge modules is better as those go into a central location, won't duplicate already existing dlls, Windows update can manage them etc. But in the short run we can either use old binary or include the runtime dll. Unless arm64 needs a new build, choose whichever is easier.

**Comment by mattock** [comment]
I agree with @rozmansi that we can cut some corners here and use the old tapinstall.exe/devcon.exe in the next installers. I'm almost certain it won't work on ARM64, but I think that is acceptable in the short-term while moving to MSI.

Am I correct that the next test installer should include everything I had in the previous one plus this PR (as-is) plus old, pre-compiled devcon.exe?

**Comment by cron2** [comment]
Hi,

On Thu, Aug 02, 2018 at 08:19:26AM +0000, Samuli Seppänen wrote:
> Am I correct that the next test installer should include everything I had in the previous one plus this PR (as-is) plus old, pre-compiled devcon.exe?

This is how i read it, yes.

gert
-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             gert@greenie.muc.de

**Comment by mattock** [comment]
I assume the "driver version now looks reasonable in device properties" is a side-effect of this PR. If so, it works as it should.

**Comment by mattock** [comment]
Test installer and signed driver for for Windows 7/8/8.1/Server 2012r2 in case somebody wants to test:
* https://build.openvpn.net/downloads/snapshots/tap-windows-9.23.1-I601.exe
* https://build.openvpn.net/downloads/snapshots/tap-windows-9.23.1.zip

**Pull request merged**